File: sdks/typescript/src/ids.ts
/**
 * Identifier helpers.
 *
 * GraphRec deduplicates events and feedback by the tenant-supplied `event_id`
 * (max 100 characters). Random IDs are fine for fire-and-forget tracking; use
 * `deterministicId` when the same business fact may be sent more than once
 * (order webhooks, replayed exports) so GraphRec records it exactly once.
 */
import { createHash, randomUUID } from "node:crypto";
import { MAX_EVENT_ID_LENGTH } from "./constants.js";

/** Random unique identifier, e.g. `evt_1f0c…`. */
export function newId(prefix = ""): string {
  const token = randomUUID().replace(/-/g, "");
  return prefix ? `${prefix}_${token}` : token;
}

export function newIdempotencyKey(): string {
  return randomUUID();
}

/**
 * Stable identifier derived from `parts` (same algorithm as the Python SDK, so
 * both SDKs produce identical IDs for identical inputs).
 *
 *     deterministicId("purchase", orderId, line, { prefix: "evt" })
 */
export function deterministicId(...args: [...parts: unknown[], options: { prefix?: string }] | unknown[]): string {
  let parts = args as unknown[];
  let prefix = "";
  const last = parts[parts.length - 1];
  if (last && typeof last === "object" && !Array.isArray(last) && "prefix" in (last as object)) {
    prefix = String((last as { prefix?: string }).prefix ?? "");
    parts = parts.slice(0, -1);
  }
  if (!parts.length) throw new RangeError("deterministicId() needs at least one part");
  const material = parts.map((part) => stringify(part)).join("\x1f");
  const token = createHash("sha256").update(material, "utf8").digest("hex").slice(0, 40);
  const result = prefix ? `${prefix}_${token}` : token;
  return result.slice(0, MAX_EVENT_ID_LENGTH);
}

/** Matches Python's `str()` for the value kinds the helpers accept. */
function stringify(part: unknown): string {
  if (part === null || part === undefined) return "None";
  if (typeof part === "boolean") return part ? "True" : "False";
  return String(part);
}

File: sdks/python/src/graphrec_sdk/_ids.py
"""Identifier helpers.

GraphRec deduplicates events and feedback by the tenant-supplied ``event_id``
(max 100 characters). Random IDs are fine for fire-and-forget tracking; use
:func:`deterministic_id` when the same business fact may be sent more than once
(order webhooks, replayed exports) so GraphRec records it exactly once.
"""

from __future__ import annotations

import hashlib
import uuid

MAX_EVENT_ID_LENGTH = 100


def new_id(prefix: str = "") -> str:
    """Random unique identifier, e.g. ``evt_1f0c...``."""

    token = uuid.uuid4().hex
    return f"{prefix}_{token}" if prefix else token


def new_idempotency_key() -> str:
    return str(uuid.uuid4())


def deterministic_id(*parts: object, prefix: str = "") -> str:
    """Stable identifier derived from ``parts``.

    >>> deterministic_id("order", 1042, "line", 2, prefix="evt")
    'evt_…'   # always the same value for the same inputs
    """

    if not parts:
        raise ValueError("deterministic_id() needs at least one part")
    material = "\x1f".join(str(part) for part in parts).encode("utf-8")
    token = hashlib.sha256(material).hexdigest()[:40]
    result = f"{prefix}_{token}" if prefix else token
    return result[:MAX_EVENT_ID_LENGTH]
